src/net_processing.cpp

```diff
@@ -3247,6 +3290,55 @@ bool PeerLogicValidation::SendMessages(CNode* pto, std::atomic<bool>& interruptM
             }
         }
 
+        // Check that outbound peers have reasonable chains
+        if (!(state.fProtectFromDisconnect || pto->fInbound || pto->m_manual_connection || pto->fFeeler || pto->fOneShot) && state.fSyncStarted) {
+            // This is an outbound peer subject to disconnection if their chain
+            // lags behind ours (note: if their chain has more work than ours,
+            // we should sync to it, unless it's invalid, in which case we
+            // should find that out and disconnect from them elsewhere).
+            if (state.pindexBestKnownBlock != nullptr && state.pindexBestKnownBlock->nChainWork >= chainActive.Tip()->nChainWork) {
+                if (state.nHeadersChainTimeout != 0) {
+                    state.nHeadersChainTimeout = 0;
+                    state.header_with_required_work = nullptr;
+                    state.fSentGetHeadersToCheckChainSync = false;
+                }
+            } else if (state.nHeadersChainTimeout == 0 || (state.header_with_required_work != nullptr && state.pindexBestKnownBlock != nullptr && state.pindexBestKnownBlock->nChainWork >= state.header_with_required_work->nChainWork)) {
```

Comment:
Style-wise, some of these statements are pretty long (this being 235 characters). This might look a bit better if split up to multi-line.